File: backend/src/services/search.service.ts
import type { MediaType, ResourceDto } from '../types/resource';
import { SearchHistoryRepository } from '../repositories/search-history.repository';
import { CloudSaverSearchService } from '../adapters/cloudsaver/cloudsaver.search';
import { PanSouSearchService } from '../adapters/pansou/pansou.search';

export class SearchService {
  constructor(
    private readonly cloudSaverSearchService: CloudSaverSearchService,
    private readonly panSouSearchService: PanSouSearchService,
    private readonly searchHistoryRepository: SearchHistoryRepository,
  ) {}

  async search(keyword: string, driver: string, mediaType: MediaType): Promise<ResourceDto[]> {
    const [cloudSaverResources, panSouResources] = await Promise.all([
      this.cloudSaverSearchService.search(keyword, driver, mediaType),
      this.panSouSearchService.search(keyword, mediaType),
    ]);

    const merged = new Map<string, ResourceDto>();
    [...cloudSaverResources, ...panSouResources].forEach((resource) => {
      const dedupeKey = `${String(resource.extra.shareCode ?? '')}:${String(resource.extra.receiveCode ?? '')}`;
      if (!merged.has(dedupeKey)) {
        merged.set(dedupeKey, resource);
      }
    });

    const resources = [...merged.values()];
    this.searchHistoryRepository.create({ keyword, driver, resultCount: resources.length });
    return resources;
  }
}

File: backend_py/app/services/search_service.py
from ..adapters.pansou import PanSouClient
from ..models.resource import MediaType, ResourceDto
from ..repositories.search_history_repository import SearchHistoryRepository


class SearchService:
    def __init__(self, pansou: PanSouClient, history: SearchHistoryRepository):
        self.pansou = pansou
        self.history = history

    async def search(self, keyword: str, driver: str, media_type: MediaType) -> list[ResourceDto]:
        resources = await self.pansou.search(keyword, media_type)
        self.history.create(keyword=keyword, driver=driver, result_count=len(resources))
        return resources
